core/src/main/java/io/undertow/conduits/DeflatingStreamSinkConduit.java

```diff
@@ -119,31 +121,35 @@ public int write(final ByteBuffer src) throws IOException {
             if (!performFlushIfRequired()) {
                 return 0;
             }
-            if (src.remaining() == 0) {
+            if (!src.hasRemaining()) {
                 return 0;
             }
-            //we may already have some input, if so compress it
-            if (!deflater.needsInput()) {
-                deflateData(false);
-                if (!deflater.needsInput()) {
-                    return 0;
-                }
-            }
```

Comment:
This case is no longer possible because we deflate from a source buffer and return the difference in position (the amount the deflater has consumed), and set an EMPTY buffer to prevent ourselves from leaking references to that buffer. As such, before and after a call to `write`, the deflater always needs input.